xdsl/interpreter.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from io import StringIO
+from typing import Any, Callable, Generator, TypeVar
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import SSAValue, Operation
+
+_OperationInvT = TypeVar('_OperationInvT', bound=Operation)
+
+
+class IntepretationError(Exception):
+    pass
+
+
+@dataclass
+class Functions:
+    functions: dict[type[Operation],
+                    Callable[[Intepreter, Operation, tuple[Any, ...]],
+                             tuple[Any, ...]]] = field(default_factory=dict)
+
+    def register_op(
+        self, op_type: type[_OperationInvT],
+        func: Callable[[Intepreter, _OperationInvT, tuple[Any, ...]],
+                       tuple[Any, ...]]):
+        self.functions[op_type] = func  # type: ignore
+
+    def op_types(self) -> set[type[Operation]]:
+        return set(self.functions.keys())
+
+    def register(
+        self, op_type: type[_OperationInvT]
+    ) -> Callable[[
+            Callable[[Intepreter, _OperationInvT, tuple[Any, ...]], tuple[Any,
+                                                                          ...]]
+    ], Callable[[Intepreter, _OperationInvT, tuple[Any, ...]], tuple[Any,
+                                                                     ...]]]:
```

Comment:
These are the type hints I am living for!

Could we make this more readable? Took me legit 3 minutes to figure out what's happening here.

Reply:
Yeah, that should work out of the box. You don't have to do anything, just using the `'O` variable will make the `TypeAlias` generic.

Reply:
Done